### ISSUE TITLE
Large RSS memory increase in MWorker when minions return after state.apply

### ISSUE THREAD

**Issue by carsten-AEI** [saltstack/salt#36819]
We currently face quite a large memory issue when running state.apply against our minions (2267 states at the moment), e.g. the resident memory size of the MWorker increase by several 100 MByte per run. When running simpler state.apply, e.g. with only 30 states, memory increase seems almost negligible, when updating 80 states it is already noticeable. We find it quite interesting, that this only happens at the very end of the run and seems to scale by the product of minion number and state numbers, when the minions return, not at the start when external pillars are evaluated etc.

We looked at various other issues w.r.t. garbage collection and tried adding gc.enable()/gc.collect() at various points of the code, but most of the time, the system will just run into the limits of RAM+swap or keeps very few GB of swap free.

```Salt Version:
           Salt: 2015.8.12

Dependency Versions:
         Jinja2: 2.7.3
       M2Crypto: Not Installed
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
         Python: 2.7.9 (default, Jun 29 2016, 13:08:31)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: 2.2
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: 1.2.3
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: 0.8.2
        timelib: Not Installed

System Versions:
           dist: debian 8.6 
        machine: x86_64
        release: 3.16.0-4-amd64
         system: debian 8.6 

```

Anything else you need from us to get an idea what may be wrong?
```

**Comment by carsten-AEI** [comment]
After quite a bit of testing, and also trying 2016.3.3 on a test system we do see vastly different RAM scaling between 2015.8.12 and 2016.3.3. Test set-up is simple a single minion and a master with 10 workers (I know overkill, but still). Then we just ran an "infinite" `state.apply` with about 2200 states from the master against the minion and tracked RAM RSS size over time via `ps`.

Plotting the min/max/avg of the workers over time yields this graph (hopefully it will show up just fine):

![salt-comparison](https://cloud.githubusercontent.com/assets/16311083/19167318/65c9556a-8c0b-11e6-9909-b51a42b8c6d7.png)

Clearly something really improved with 2016.3 - I am not sure what but as the RAM usage increases in 2015.8 b ya lot when the minions return, I could imagine that the fairly large replies are somehow kept in memory for too long.

Bottom line: 2016.3.3 seems to fix our immediate problem, large scale testing will commence tomorrow, so not 100% sure what to do with this issue.

**Comment by Ch3LL** [comment]
@carsten-AEI is there anything unique about your environment? you mention external pillars do you happen to be using git pillars? Would you mind sharing a sanitized version of your master/minion to see if what is unique in your configs. 

Also let us know how the 2016.3.3 testing goes :)

**Comment by carsten-AEI** [comment]
Hi @Ch3LL, sorry I forgot to update our finding after today's upgrade.

Not sure if our set-up is normal or quite different, too new to salt to be sure. Changes to master conf:
```
keep_jobs: 6
timeout: 30
loop_interval: 300
con_cache: True
worker_threads: 250
pub_hwm: 10000
salt_event_pub_hwm: 64000
event_publisher_pub_hwm: 32000
state_output: changes
state_verbose: False
state_aggregate: True
file_roots:
  base:
    - /srv/salt/master
hash_type: sha256
pillar_roots:
  base:
    - /srv/salt/pillar
log_level: info
```

We use several pillars, regular, external-file-tree as well as external-sqlite3 with four queries.

Master machine is mid sized box is a quad-core 3.3GHz E3-1230v2 (HT enabled), 32 GB box.

After the upgrade, we were able to run `salt -N compute-nodes -b 50 state.apply` without a problem - after an hour all (at that time) 20 workers were pretty busy and after an hour all 2100 minions got their 2169 states updates. All workers were using at most 80MB RSS, only the command line client `salt` was using 10GB in the end.

After that we investigated scaling a bit more to see if we could push out the updates faster, but we had to resort to large timeout values or smaller batch sizes, e.g. `-b 500 -t 180` went ok, `-b 500 -t 60` did not with a typical error like:

```
2016-10-07 12:00:14,870 [salt.loaded.int.output.highstate][ERROR   ][24248] Data passed to highstate outputter is not a valid highstate return: {'a9521.atlas.local': ['Traceback 
(most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/state.py", line 3102, in call_highstate
    top = self.get_top()
  File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2589, in get_top
    tops = self.get_tops()
  File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2438, in get_tops
    for saltenv in self._get_envs():
  File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2363, in _get_envs
    client_envs = self.client.envs()
  File "/usr/lib/python2.7/dist-packages/salt/fileclient.py", line 1327, in envs
    return self.channel.send(load)
  File "/usr/lib/python2.7/dist-packages/salt/utils/async.py", line 74, in wrap
    ret = self._block_future(ret)
  File "/usr/lib/python2.7/dist-packages/salt/utils/async.py", line 84, in _block_future
    return future.result()
  File "/usr/lib/python2.7/dist-packages/tornado/concurrent.py", line 215, in result
    raise_exc_info(self._exc_info)
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 876, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib/python2.7/dist-packages/salt/transport/zeromq.py", line 240, in send
    ret = yield self._crypted_transfer(load, tries=tries, timeout=timeout)
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 870, in run
    value = future.result()
  File "/usr/lib/python2.7/dist-packages/tornado/concurrent.py", line 215, in result
    raise_exc_info(self._exc_info)
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 876, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib/python2.7/dist-packages/salt/transport/zeromq.py", line 208, in _crypted_transfer
    ret = yield _do_transfer()
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 870, in run
    value = future.result()
  File "/usr/lib/python2.7/dist-packages/tornado/concurrent.py", line 215, in result
    raise_exc_info(self._exc_info)
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 876, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib/python2.7/dist-packages/salt/transport/zeromq.py", line 194, in _do_transfer
    tries=tries,
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 870, in run
    value = future.result()
  File "/usr/lib/python2.7/dist-packages/tornado/concurrent.py", line 215, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 3, in raise_exc_info
SaltReqTimeoutError: Message timed out'
]}
```

(please note that at the moment the minions are still on 2015.8.12)

So, tl;dr - the memory problem does seem to be solved by upgrading the master to 2016.3.3, next up is trying to get updates out faster - a large time-out value is a bit unfortunate, as `salt -b X` will wait until the time-out is reached... ;)

**Comment by Ch3LL** [comment]
@carsten-AEI glad you got the memory issue working :)

ping @cachedout any ideas on how to get out the updates quicker in @carsten-AEI 's setup?

But also another question as well: are all your minions 2016.3.3 as well?

**Comment by carsten-AEI** [comment]
Sorry for lack of updates - I currently try to sneak in a few vacation days. I guess we will try to update minions to 2016.3 within the next two weeks and see if that will change timings even more.

**Comment by carsten-AEI** [comment]
Sorry for the long delay, but after a month of testing, memory usage is very stable on 2016.3 and thus I think we can tick this issue off as solved (at least for us). We are still struggling with getting a higher concurrency in updating > 2000 minions, but at the moment it looks like the machine is just too small for that.

**Comment by Ch3LL** [comment]
Perfect! You okay to close this and then open a new issue if you find out that resources is not the issue for concurrency issue?

**Comment by carsten-AEI** [comment]
Yes, let's close this one. It has been stable over the past few months. Thanks!
